File: zenscript-language-server/src/main/java/raylras/zen/lsp/provider/DocumentSymbolProvider.java
package raylras.zen.lsp.provider;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.RuleNode;
import org.eclipse.lsp4j.DocumentSymbol;
import org.eclipse.lsp4j.DocumentSymbolParams;
import org.eclipse.lsp4j.Range;
import org.eclipse.lsp4j.SymbolKind;
import raylras.zen.model.CompilationUnit;
import raylras.zen.model.Visitor;
import raylras.zen.model.parser.ZenScriptParser.*;
import raylras.zen.util.ArrayStack;
import raylras.zen.util.CSTNodes;
import raylras.zen.util.Ranges;
import raylras.zen.util.Stack;

import java.util.ArrayList;
import java.util.List;

public final class DocumentSymbolProvider {

    private DocumentSymbolProvider() {}

    public static List<DocumentSymbol> documentSymbol(CompilationUnit unit, DocumentSymbolParams params) {
        DocumentSymbolVisitor visitor = new DocumentSymbolVisitor();
        unit.getParseTree().accept(visitor);
        return visitor.topLevelSymbolList;
    }

    private static final class DocumentSymbolVisitor extends Visitor<DocumentSymbol> {
        private final List<DocumentSymbol> topLevelSymbolList = new ArrayList<>();
        private final Stack<DocumentSymbol> stack = new ArrayStack<>();

        @Override
        public DocumentSymbol visitFunctionDeclaration(FunctionDeclarationContext ctx) {
            return enter(ctx, ctx.simpleName(), SymbolKind.Function);
        }

        @Override
        public DocumentSymbol visitClassDeclaration(ClassDeclarationContext ctx) {
            return enter(ctx, ctx.simpleNameOrPrimitiveType(), SymbolKind.Class);
        }

        @Override
        public DocumentSymbol visitExpandFunctionDeclaration(ExpandFunctionDeclarationContext ctx) {
            return enter(ctx, ctx.simpleName(), SymbolKind.Function);
        }

        @Override
        public DocumentSymbol visitConstructorDeclaration(ConstructorDeclarationContext ctx) {
            return enter(ctx, ctx.ZEN_CONSTRUCTOR(), SymbolKind.Constructor);
        }

        @Override
        public DocumentSymbol visitVariableDeclaration(VariableDeclarationContext ctx) {
            return enter(ctx, ctx.simpleName(), SymbolKind.Variable);
        }

        @Override
        public DocumentSymbol visitFunctionExpr(FunctionExprContext ctx) {
            return enter(ctx, ctx.FUNCTION(), SymbolKind.Function);
        }

        private DocumentSymbol enter(RuleNode enclose, ParseTree selection, SymbolKind kind) {
            DocumentSymbol symbol = toDocumentSymbol(enclose, selection, kind);
            if (push(symbol)) {
                visitChildren(enclose);
                pop();
            }
            return symbol;
        }

        private DocumentSymbol toDocumentSymbol(ParseTree enclose, ParseTree selection, SymbolKind kind) {
            return toDocumentSymbol(enclose, selection, CSTNodes.getText(selection), kind);
        }

        private DocumentSymbol toDocumentSymbol(ParseTree enclose, ParseTree selection, String name, SymbolKind kind) {
            Range encloseRange = Ranges.toLspRange(enclose);
            Range selectionRange = Ranges.toLspRange(selection);
            return new DocumentSymbol(name, kind, encloseRange, selectionRange);
        }

        private boolean push(DocumentSymbol symbol) {
            if (symbol.getName().isEmpty()) {
                return false;
            }
            if (isTopLevel()) {
                addToTopLevelSymbolList(symbol);
            } else {
                addToCurrentSymbolChildren(symbol);
            }
            stack.push(symbol);
            return true;
        }

        private void pop() {
            stack.pop();
        }

        private boolean isTopLevel() {
            return stack.isEmpty();
        }

        private void addToTopLevelSymbolList(DocumentSymbol symbol) {
            topLevelSymbolList.add(symbol);
        }

        private void addToCurrentSymbolChildren(DocumentSymbol symbol) {
            DocumentSymbol parent = stack.peek();
            if (parent.getChildren() == null) {
                parent.setChildren(new ArrayList<>());
            }
            parent.getChildren().add(symbol);
        }
    }

}

File: zenscript-language-server/src/main/kotlin/raylras/zen/lsp/provider/DocumentSymbolProvider.kt
package raylras.zen.lsp.provider

import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.RuleNode
import org.eclipse.lsp4j.DocumentSymbol
import org.eclipse.lsp4j.DocumentSymbolParams
import org.eclipse.lsp4j.SymbolKind
import raylras.zen.model.CompilationUnit
import raylras.zen.model.Visitor
import raylras.zen.model.parser.ZenScriptParser.*
import raylras.zen.util.ArrayStack
import raylras.zen.util.textRange
import raylras.zen.util.toLspRange

object DocumentSymbolProvider {
    fun documentSymbol(unit: CompilationUnit, params: DocumentSymbolParams): List<DocumentSymbol> {
        val visitor = DocumentSymbolVisitor()
        unit.parseTree!!.accept(visitor)
        return visitor.topLevelSymbolList
    }

    private class DocumentSymbolVisitor : Visitor<Unit>() {
        val topLevelSymbolList = ArrayList<DocumentSymbol>()
        private val stack = ArrayStack<DocumentSymbol>()

        override fun visitFunctionDeclaration(ctx: FunctionDeclarationContext) {
            enterSymbol(ctx, ctx.simpleName(), SymbolKind.Function)
        }

        override fun visitClassDeclaration(ctx: ClassDeclarationContext) {
            enterSymbol(ctx, ctx.simpleName() ?: ctx.simpleNameOrPrimitiveType(), SymbolKind.Class)
        }

        override fun visitExpandFunctionDeclaration(ctx: ExpandFunctionDeclarationContext) {
            enterSymbol(ctx, ctx.simpleName(), SymbolKind.Function)
        }

        override fun visitConstructorDeclaration(ctx: ConstructorDeclarationContext) {
            enterSymbol(ctx, ctx.ZEN_CONSTRUCTOR(), SymbolKind.Constructor)
        }

        override fun visitVariableDeclaration(ctx: VariableDeclarationContext) {
            enterSymbol(ctx, ctx.simpleName(), SymbolKind.Variable)
        }

        override fun visitFunctionExpr(ctx: FunctionExprContext) {
            enterSymbol(ctx, ctx.FUNCTION(), SymbolKind.Function)
        }

        private fun enterSymbol(
            encloseCtx: RuleNode,
            selectionCtx: ParseTree?,
            kind: SymbolKind,
            name: String? = selectionCtx?.text
        ) {
            selectionCtx ?: return
            name ?: return
            toDocumentSymbol(encloseCtx, selectionCtx, name, kind).let {
                push(it)
                visitChildren(encloseCtx)
                pop()
            }
        }

        private fun toDocumentSymbol(
            encloseCtx: ParseTree,
            selectionCtx: ParseTree,
            name: String,
            kind: SymbolKind
        ): DocumentSymbol {
            val encloseRange = encloseCtx.textRange.toLspRange()
            val selectionRange = selectionCtx.textRange.toLspRange()
            return DocumentSymbol(name, kind, encloseRange, selectionRange)
        }

        private fun push(symbol: DocumentSymbol) {
            if (isTopLevel) {
                addToTopLevelSymbolList(symbol)
            } else {
                addToCurrentSymbolChildren(symbol)
            }
            stack.push(symbol)
        }

        private fun pop() {
            stack.pop()
        }

        private val isTopLevel: Boolean
            get() = stack.isEmpty()

        private fun addToTopLevelSymbolList(symbol: DocumentSymbol) {
            topLevelSymbolList.add(symbol)
        }

        private fun addToCurrentSymbolChildren(symbol: DocumentSymbol) {
            val parent = stack.peek() ?: return
            if (parent.children == null) {
                parent.children = ArrayList()
            }
            parent.children.add(symbol)
        }
    }
}
